### DISCUSSION TITLE
Is it possible to provide a Lint plugin?

### DISCUSSION THREAD

**Discussion by rakuyoMo** [airbnb/swift#269] [Ideas]
SwiftLint provides a [SwiftLintBuildToolPlugin](https://github.com/realm/SwiftLint/blob/5315c3d1b68ebae7bf3a5871c22e096d67c32e6e/Package.swift#L22C24-L22C48) plug-in for real-time prompting of Lint in Xcode.

I noticed that AirbnbSwift does not seem to provide such a plug-in at present. Do all projects within the Airbnb organization rely on the command line to view Lint results?

Is it possible to provide a plug-in similar to SwiftLintBuildToolPlugin to report Lint content in real time?

**Comment by rakuyoMo** [comment]
I feel like this is important to me, but I don't seem to have the ability to implement it... 

When I bring in Airbnb/swift via an SPM plugin, I don't know the path to the `swiftlint.yml` file (and whether it will change dynamically) , so I have no way to add scripts in build phases. In addition, when I focus on developing an SPM package without a project, I have no way to view lint warnings in real time.

**Reply by rakuyoMo** [reply]
I also have no way to add `Target.PluginUsage.plugin(name: "FormatSwift", package: "swift")` in `Package.swift` because it does not support passing parameters.

**Comment by calda** [comment]
Currently within Airbnb, engineers run the code formatter / linter via the command line and see any output there. 

I think it would be nice for SwiftLint violations to show up in Xcode! A build tool plugin would work nicely for this. If you are interested in writing one of these, we'd certainly be open to accepting it as a part of this repo via a PR!

On the other hand, the majority of our rules are implemented via SwiftFormat. While SwiftFormat _can_ be used as a linter, it's really a much better experience when used as a code formatter (since it automatically fixes your code for you). I'm very interested in having an Xcode code editor plugin that supports running SwiftFormat and SwiftLint autocorrect directly in the Xcode editor using our formatting configuration. We're exploring this internally, and that work may make it over to this open source repo (depending on if it winds up having a lot of internal dependencies).

**Reply by rakuyoMo** [reply]
Is the plug-in you are making now similar to [SwiftFormat](https://github.com/nicklockwood/SwiftFormat/blob/825b4cdadbcabb91661243a09aeb44c3d4ef459d/Package%40swift-5.6.swift#L15)? Because I found that AirbnbFormatTool seems not to be used directly on Xcode.

**Reply by calda** [reply]
This repo does provide an Xcode command plugin: https://github.com/airbnb/swift/blob/master/Plugins/FormatSwift/Plugin.swift#L139. You can invoke it via the Xcode UI. I don't personally think it's a great experience though, since it requires clicking through a few menus every time.

Xcode source editor extensions provide a better experience imo. SwiftFormat publishes one here: https://github.com/nicklockwood/SwiftFormat?tab=readme-ov-file#xcode-source-editor-extension
